plugins/ebt/loop.go

```diff
@@ -0,0 +1,211 @@
+				}
+
+				err := json.Unmarshal(body, &msgWithAuthor)
+				if err != nil {
+					h.check(fmt.Errorf("unable to establish author for format %s: %w", format, err))
+					continue
+				}
+				author = msgWithAuthor.Author
+
+			case refs.RefAlgoFeedBendyButt:
+				var msg metafeed.Message
+				err = msg.UnmarshalBencode(body)
+				if err != nil {
+					h.check(fmt.Errorf("unable to establish authorfor format %s: %w", format, err))
+					continue
+				}
+
+				author = msg.Author()
+
+			case refs.RefAlgoFeedGabby:
+				var msg gabbygrove.Transfer
+				err = msg.UnmarshalCBOR(body)
+				if err != nil {
+					h.check(fmt.Errorf("unable to establish authorfor format %s: %w", format, err))
+					continue
+				}
+
+				author = msg.Author()
+
+			default:
+				h.check(fmt.Errorf("unhandled format: %s", format))
+				return
+			}
+
+			vsnk, err := h.verify.GetSink(author, true)
+			if err != nil {
+				h.check(err)
+				continue
+			}
+
+			err = vsnk.Verify(body)
+			if err != nil {
+				// TODO: mark feed as bad
+				h.check(err)
+			}
+
+			continue
+		}
+
+		// update our network perception
+		wants, err := h.stateMatrix.Update(peer, frontierUpdate)
+		if err != nil {
+			h.check(err)
+			return
+		}
+
+		// TODO: partition wants across the open connections
+		// one peer might be closer to a feed
+		// for this we also need timing and other heuristics
+
+		// ad-hoc send where we have newer messages
+		for feedStr, their := range wants.Frontier {
+			// these were already validated by the .UnmarshalJSON() method
+			// but we need the refs.Feed for the createHistArgs
+			feed, err := refs.ParseFeedRef(feedStr)
+			if err != nil {
+				h.check(err)
+				return
+			}
+
+			// skip feeds not in the desired format for this sessions
+			if feed.Algo() != format {
+				continue
+			}
+
+			if !their.Replicate {
+				continue
+			}
+
+			if !their.Receive {
+				session.Unubscribe(feed)
+				continue
+			}
+
+			arg := message.CreateHistArgs{
+				ID:  feed,
+				Seq: int64(their.Seq + 1),
+			}
+			arg.Limit = -1
+			arg.Live = true
+
+			// TODO: it might not scale to do this with contexts (each one has a goroutine)
+			// in that case we need to rework the internal/luigiutils MultiSink so that we can unsubscribe on it directly
+			ctx, cancel := context.WithCancel(ctx)
+
+			err = h.livefeeds.CreateStreamHistory(ctx, tx, arg)
+			if err != nil {
+				cancel()
+				h.check(err)
+				return
+			}
+			session.Subscribed(feed, cancel)
+		}
+	}
+
+	h.check(rx.Err())
+}
+
+func (h *MUXRPCHandler) sendState(ctx context.Context, tx *muxrpc.ByteSink, remote refs.FeedRef) error {
+	currState, err := h.stateMatrix.Changed(h.self, remote)
+	if err != nil {
+		return fmt.Errorf("failed to get changed frontier: %w", err)
+	}
+
+	selfRef := h.self.Ref()
+
+	// don't receive your own feed
+	if myNote, has := currState.Frontier[selfRef]; has {
+		myNote.Receive = false
+		currState.Frontier[selfRef] = myNote
+	}
+
+	tx.SetEncoding(muxrpc.TypeJSON)
```

Comment:
I couldn't really decide what to do about the state/frontier transmissions in binary mode so i kept them in JSON.

I'm not sure if the JS stack is fine with `[JSON, binary, binary, binary, JSON, binary, ...]` framed muxrpc streams, tho.

cc @arj03

Reply:
Thanks for the ping @cryptix. I'll have a careful look at this and get back.

Reply:
bencode and cbor would be fine transmitting maps, so we can solve it either way IMHO, this was just the easiest for me.

Reply:
I believe this `[JSON, binary, binary, binary, JSON, binary, ...] ` is how it is done now if you replicate binary data